metadata/src/from/v14.rs

```diff
@@ -0,0 +1,313 @@
+                    })
+                    .collect(),
+            });
+            let constants = p.constants.into_iter().map(|c| {
+                let name: ArcStr = c.name.clone().into();
+                (name.clone(), from_constant_metadata(name, c))
+            });
+
+            let call_variant_index =
+                VariantIndex::build(p.calls.as_ref().map(|c| c.ty.id), &m.types);
+            let error_variant_index =
+                VariantIndex::build(p.error.as_ref().map(|e| e.ty.id), &m.types);
+            let event_variant_index =
+                VariantIndex::build(p.event.as_ref().map(|e| e.ty.id), &m.types);
+
+            pallets_by_index.insert(p.index, pos);
+            pallets.push_insert(
+                name.clone(),
+                PalletMetadataInner {
+                    name,
+                    index: p.index,
+                    storage,
+                    call_ty: p.calls.map(|c| c.ty.id),
+                    call_variant_index,
+                    event_ty: p.event.map(|e| e.ty.id),
+                    event_variant_index,
+                    error_ty: p.error.map(|e| e.ty.id),
+                    error_variant_index,
+                    constants: constants.collect(),
+                    view_functions: vec![],
+                    associated_types: Default::default(),
+                    docs: vec![],
+                },
+            );
+        }
+
+        let dispatch_error_ty = m
+            .types
+            .types
+            .iter()
+            .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
+            .map(|ty| ty.id);
+
+        Ok(Metadata {
+            types: m.types,
+            pallets,
+            pallets_by_index,
+            extrinsic: from_extrinsic_metadata(m.extrinsic, missing_extrinsic_type_ids),
+            dispatch_error_ty,
+            outer_enums: OuterEnumsMetadata {
+                call_enum_ty: outer_enums.call_enum_ty.id,
+                event_enum_ty: outer_enums.event_enum_ty.id,
+                error_enum_ty: outer_enums.error_enum_ty.id,
+            },
+            apis: Default::default(),
+            custom: CustomMetadataInner {
+                map: Default::default(),
+            },
+        })
+    }
+}
+
+fn from_signed_extension_metadata(
+    value: v14::SignedExtensionMetadata<PortableForm>,
+) -> TransactionExtensionMetadataInner {
+    TransactionExtensionMetadataInner {
+        identifier: value.identifier,
+        extra_ty: value.ty.id,
+        additional_ty: value.additional_signed.id,
+    }
+}
+
+fn from_extrinsic_metadata(
+    value: v14::ExtrinsicMetadata<PortableForm>,
+    missing_ids: MissingExtrinsicTypeIds,
+) -> ExtrinsicMetadata {
+    let transaction_extensions: Vec<_> = value
+        .signed_extensions
+        .into_iter()
+        .map(from_signed_extension_metadata)
+        .collect();
+
+    let transaction_extension_indexes = (0..transaction_extensions.len() as u32).collect();
```

Comment:
I wonder whether we can replace Vec<u32> with Range<u32> as transaction_extensions_indexes instead to get rid off some alloc?

Reply:
In this case we could, but for v16 metadata, the indexes vec might no longer be contiguous if we have several versions of transaction extensions defined.

Reply:
I see